### PR TITLE
feat(journal): add assert for concurrent programs

### PR DESCRIPTION
**Pull request by symbiont-stevan-andjelkovic**

### CODE REVIEW

File: src/journal/test/JournalTest.hs [hunk 1]
```diff
@@ -112,14 +112,16 @@ appendBSFake bs fj@(FakeJournal bss ix termCount) =
     hardLimit = readBytes + termLen -- TODO: calculate this correctly
 
     readBytes :: Int
-    readBytes = sum [ align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT
-                    | bs <- map (bss Vector.!) [0..ix - 1]
-                    ]
+    readBytes = Vector.sum
+              . Vector.map (\bs -> align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT)
+              . Vector.slice 0 (max 0 (ix - 1))
```
Comment (symbiont-daniel-gustafsson) [inline src/journal/test/JournalTest.hs:117]: Is this `Vector.take (ix - 1)`?

File: src/journal/test/JournalTest.hs [hunk 2]
```diff
@@ -112,14 +112,16 @@ appendBSFake bs fj@(FakeJournal bss ix termCount) =
...
+              $ bss
 
     unreadBytes :: Int
-    unreadBytes = sum [ align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT
-                      | bs <- map (bss Vector.!) [ix..Vector.length bss - 1]
-                      ]
+    unreadBytes = Vector.sum
+                . Vector.map (\bs -> align (hEADER_LENGTH + BS.length bs) fRAME_ALIGNMENT)
+                . Vector.slice ix (max 0 (Vector.length bss - 1 - ix))
```
Comment (symbiont-daniel-gustafsson) [inline src/journal/test/JournalTest.hs:123]: Is this `Vector.drop ix`?

File: src/journal/test/JournalTest.hs [hunk 3]
```diff
@@ -742,7 +757,7 @@ genHistory pids initModel = sized $ \s ->
           go (consH (Ok pid resp) conc) linear model size ((pid, DoingNothing):pids')
 
 prop_lineariseIsOkay :: Property
-prop_lineariseIsOkay = mapSize (max 20) $
+prop_lineariseIsOkay = mapSize (min 20) $
```
Comment (symbiont-daniel-gustafsson) [inline src/journal/test/JournalTest.hs:760]: I don't think you want `min 20` here it might just take more time than needed.

**Review by symbiont-daniel-gustafsson — APPROVED** [review]

File: src/journal/test/JournalTest.hs (hunk 3 above)
Comment (symbiont-stevan-andjelkovic) [reply src/journal/test/JournalTest.hs:760]: I think `max 20` was wrong if we want same semantics as `withMaxSize 20`? `min 20` will allow smaller sizes but cap the size at 20.

File: src/journal/test/JournalTest.hs (hunk 1 above)
Comment (symbiont-stevan-andjelkovic) [reply src/journal/test/JournalTest.hs:117]: True, that's much simpler, thanks!

File: src/journal/test/JournalTest.hs (hunk 3 above)
Comment (symbiont-daniel-gustafsson) [reply src/journal/test/JournalTest.hs:760]: Oh you are right

**Pull request merged**